README
                      Breakaway 2.1 -- General Information
October 2014

Breakaway 2.1 is an application that allows you to automatically play/pause
iTunes by using the headphone jack and/or the mute button. It is similar to the
built-in functionality offered by an iPod.

                             - - - - - - - - - - -

                                  Basic Usage

The usage of Breakaway is very straightforward. Simply have Breakaway and iTunes
running at the same time, then preform one of the following actions:
    o Press the mute button
    o Connect/disconnect headphones from the headphones jack

Depending on the playing status of iTunes, Breakaway will intelligently decide
what to do. For example, consider iTunes playing and the headphones connected.
If the headphones are suddenly disconnected, Breakaway would pause iTunes. If
they are connected again, Breakaway would resume play. Similar scenarios exist
with the mute button.

Breakaway will always try to follow your lead. If you set iTunes to play,
Breakaway will honor that state and always revert back to it whenever possible.
Similarly, if you set iTunes to pause, Breakaway will not attempt to set it to
play.

                             - - - - - - - - - - -

                                  Preferences

General
    Automatically start at login: Breakaway will be added to your Login Items
    (manually accessable via System Preferences->Accounts).

    Show in menu bar: A menu extra will be available in the top right of your
    screen for quick access and mode display.

    Show in Dock: Requires a restart for changes to take effect. Shows
    Breakaway's icon in the Dock.

    Fade-in: Controls how fast the fade-in happens. Moving the slider all the
    way to the right turns the feature off.

Plugins
    The left table houses all plugins installed and loaded. Selecting an item on
    the left reveals it's contents on the right table. If the plugin is designed
    to be created multiple times, you can add/remove instances of the plugin
    with the +/- buttons at the bottom of the screen. Plugins' options are found
    by opening the info drawer (the "i" button). 

    If you want to add/delete plugins, click on the button under the left table.
    This will open a window to the plugin directory. Simply add/delete plugins
    like you would any other file: drag it to Trash.
    
Expand Breakaway
    Enable anonymous system profiling: As stated under the preference, all
    information is anonymously sent. Specifically, the information being sent
    is:
        o Mac OS X version
        o CPU type/subtype
        o Mac model
        o Number of CPUs
        o 32-bit vs. 64-bit
        o CPU speed
        o RAM size
        o Application name
        o Application version
        o User's preferred language
    
Advanced System Diagnostics
    Follow the onscreen directions. This window can be used to provide more
    anonymous data that could not be collected during the former profiling
    because it requires user interaction.

    This information is more useful for debugging and seeing if Breakaway is
    correctly operating on the machine in question. If you are experiencing
    issues, use this window to generate a report of Breakaway's main functions
    and send it to the developer. It takes about 5 seconds to complete.

Update
    Automatically check for updates: On startup, Breakaway will check for new
    versions and notify the user if there is one available.

    Automatically download updates: If there is a download available, Breakaway
    will automatically download it and prompt the user once it is ready for
    install.

                             - - - - - - - - - - -

                           Frequently Asked Questions

Breakaway does not work.
    * Make sure that Breakaway is enabled (through menu extra).
    * Make sure that you are not using any fancy audio setups. If you are,
      please contact the developer to see if there is a possible solution.

Breakaway is acting strange or not operating properly.
    Try the following solutions (one at a time):
        * Start Breakaway (if is not running already)
            o Quit iTunes
            o Start iTunes

        * Quit Breakaway
            o Quit iTunes
            o Start Breakaway
            o Start iTunes

        * Quit Breakaway
            o Move the preference file
              ([~]/Library/Preferences/com.mutablecode.breakaway.plist) to the
              Desktop
            o Start Breakaway

How do I install Breakaway?
    Drag the "Breakaway.app" file anywhere you like.

How do I uninstall Breakaway?
    * Place the "Breakaway.app" file in Trash
    * Place the preference file "com.mutablecode.breakaway.plist" in Trash
      (located in [~]/Library/Preferences/)
    * If the directory [~]/Library/Application Support/Breakaway/ exists, place
      it in Trash

Does Breakaway support any other application besides iTunes?
    No, Breakaway only works with iTunes. Breakaway does support plugins though,
    if you want to try.

Can I help with programming or localizing?
    Sure, I would love the extra help. Just send me an email and we can discuss
    it.

Do you (Kevin) accept donations?
    Graciously.
                             - - - - - - - - - - -

                               Support Contact

               https://github.com/steakknife/breakaway/issues

                             - - - - - - - - - - -

                   Original Author (don't contact for support)

Name: Kevin Nygaard
Email: admin@mutablecode.com

                             - - - - - - - - - - -

                                    License

Breakaway is free software: you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.

Breakaway is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
Breakaway.  If not, see <http://www.gnu.org/licenses/>.


ChangeLog
Developer's Log
Kevin Nygaard | March 2010

0.1: Proof of concept through Apple's HALLab
     Programming commences
     Program complete but with lag in callback Applescript object
     Bug in callback severely hamstrings functionality

0.2: Lag solved by kicking callback on start
     Bug in callback resolved. Functionality restored

0.3: Added preferences  - Mute Key Functionality
            - Resume play on reinsert
            - Force open iTunes
     Made application run smoother

0.4: Added preferences  - Allow initial compile
            - Use Type B
     Application is fully functional now

0.5: Optimized iTunesActive code
     First public release

0.6: Fixed potentially fatal Type B declaration (IntelMac problem only)
     Added menu extra disable functionality
     Preference window to front on app icon click
     Added splash
     New icon
     Halved menu extra icon size
     Added quit button to preferences
     
0.7: Added Growl support
     Fixed mute functionality on G5/Intel iMacs
     Fixed minor mute key functionality to headphones mode
     Added SmartPlay feature to normal mode
     Made splash disappear quicker
     Splash now displays correct version
     Miscellaneous memory management enhancements
     Fixed potential "Force iTunes Open" problem
     Added about button to preferences
     Preferences now come to active front when called for 

0.8: "Resume on Reinsert" option now extends to Mute functionality
     Removed "Initial Compile" option (no more need for it)
     "Force-Open iTunes" now functions properly
     Memory Management
     Much faster response times

0.9: Added more to SmartPlay (will not resume iTunes if there is another audio source playing)
     Breakaway will now only start playing music if it stopped it (changeable via hidden preference)
     Breakaway will now only start playing music if there is no other other audio source playing (disabled by default | see readme)
     Enable/Disable bug fixed
     Preferences improved
     Mute functionality improved
     Mute property listener fixed

0.95: Added catch should mute/headphones be pressed/pulled in quick succession
      Should correctly function with MacBooks (still not MacBook Pro)

1.0: Now works with Macbook Pros and Mac Pros
     Fixed error when enabling/disabling

1.1: Preference window can now be closed with expected hotkeys (Command+W)
     New preference attempts to guess what mode you want to be in

1.2: New preference window
     New preferences: Show in Dock
                      Proc throwout time
     New icons
     New Triggers functionality
     New auto updating with Sparkle
     AppleScripts accepted for triggers 
     ApplesScripts are now thread safe
     Mute key functionality now defaults to on
     Fixed potential problem with the Disable system
     Fixed problem where Disable system would not fully disable itself
	 
1.3: Fixed error in auto update where it would not follow the user preference
     Triggers now are more useful (activate regardless of what iTunes is doing)
		Pseudo-VLC support (somewhat slow in comparison to iTunes)
		Pseudo-Anything-that-will-take-AppleScript support
	 Triggers are optimized to a higher degree

1.4: Preference to turn off splash image

1.5: Bug where Breakaway would not proc from 0 to +

1.6: Chinese (Simplified) Localization
     Dutch Localization
     Fixed error on first start where auto update window is behind splash
     Fixed error where Breakaway would get utterly confused if you ran it when iTunes was playing
     Major bug where debug settings were included in release version, resulting in very very poor memory management
     "Resume on reinsert" now defaults to on
     Added fade-in options
     Added "Expand Breakaway" system support feedback option

1.7: Fixed minor UI errors where auto updater and info panel appeared beneath active window
     Fixed bug where Breakaway would incorrectly read mute status when plugging/unplugging headphones
     Fixed many logic errors concerning triggers
     Modified trigger interface
     Updated Chinese (Simplified) localization
     Updated French localization
     Minor UI adjustments

1.75b: [New] Now open source on SourceForge.net
       [New] Now under GPL license
	   [New] Plugins added
	   [Change] Triggers replaced with Plugins
	   [Change] iTunes playpause calls now significantly faster
	   [Change] Preferences rearranged

1.8: [NEW] Optional anonymous system profiling via Sparkle
	 [FIX] Much more reliable in general
	 [FIX] Fade-in works better
	 [FIX] Fixed error where the status bar image would change when breakaway is in disabled mode
	 [FIX] Show dock icon works better now
	 [CHANGE] Mute key always watched now
	 [CHANGE] "Guess mode" option always on
	 [CHANGE] "SmartPlay" option always on
	 [CHANGE] Preferences redesigned
	 [CHANGE] Spalsh image removed
	 [CHANGE] "Force open" option removed
	 [CHANGE] "Audio flowing" feature removed

1.9: [FIX] Menu item status icon bug fixed
     [FIX] Corrected a few errors regarding Advanced Diagnostics
     [FIX] General stability and efficiency fixes
     [CHANGE] Breakaway respects iTunes volume now during fade-in
     [CHANGE] Fade-in preferences redesigned

2.0: [NEW] Menubar icon is now animated
	 [NEW] New, simpler interface
	 [NEW] Operates entirely on plugin-based architecture
	 [NEW] Applescript plugins are based in CoreData
	 [FIX] Applescript plugins work with iTunes not running
	 [FIX] Lots of optimizations

2.0.1: [FIX] Preferences accessible through icon double-click

2.1.0: [NEW] Update for Yosemite
       [BUG] Enable isn't currently respected
	 
===========================================================================


AITrigger.m
/*
 * AITrigger.m
 * Breakaway
 * Created by Kevin Nygaard on 8/15/07.
 * Copyright 2008 Kevin Nygaard.
 *
 * This file is part of Breakaway.
 *
 * Breakaway is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Breakaway is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Breakaway.  If not, see <http://www.gnu.org/licenses/>.
 */

#import "AITrigger.h"

#define HEADPHONES_MODE 2
#define NORMAL_MODE 4
#define MUTE 8
#define UNMUTE 16
#define HEADPHONES_IN 32
#define HEADPHONES_OUT 64

@implementation AITrigger
-(id)initFromDictionary:(NSDictionary*)attributes
{
	if (!(self = [super init])) return nil;
	
	//if ([attributes count] < 10) { NSLog(@"Not enough attributes to pop. Returning NULL."); return NULL; }
	
    [self setName:[attributes objectForKey:@"name"]];
	
	// Modes
    [self setnMode:[[attributes objectForKey:@"nmode"]boolValue]];
	[self sethpMode:[[attributes objectForKey:@"hpmode"]boolValue]];
	
	// Triggers
	[self setMute:[[attributes objectForKey:@"mute"]boolValue]];
	[self setUnmute:[[attributes objectForKey:@"unmute"]boolValue]];
	[self setHin:[[attributes objectForKey:@"hin"]boolValue]];
	[self setHout:[[attributes objectForKey:@"hout"]boolValue]];
	
    [self setLod:[[attributes objectForKey:@"lod"]intValue]];
	[self setScript:[attributes objectForKey:@"script"]];
	[self setEnabled:[[attributes objectForKey:@"enabled"]boolValue]];
	
	[self setFamilyCode];
	
	// if we aren't loading on demand, compile now and be done with it
	if (!lod) [self compile];
	
	return self;
}

-(void)compile
{
	if([applescript source]== nil) [self setScript:nil];
	[applescript compileAndReturnError:nil];
}

-(void)execute
{
	NSLog(@"::Launching Script::\n::%@::\n::%@::",[applescript source],[applescript isCompiled]?@"COMPILED":@"NOT COMPILED");
	if (![applescript isCompiled]) [applescript performSelectorOnMainThread:@selector(compileAndReturnError:) withObject:nil waitUntilDone:NO];
	[applescript performSelectorOnMainThread:@selector(executeAndReturnError:) withObject:nil waitUntilDone:NO];
}

#pragma mark KVC Accessors
//{{{ KVC Functions
-(NSString*)name
{
	return name;
}

-(BOOL)nmode
{
	return nmode;
}

-(BOOL)hpmode
{
	return hpmode;
}

-(BOOL)mute
{
	return mute;
}

-(BOOL)unmute
{
	return unmute;
}

-(BOOL)hin
{
	return hin;
}

-(BOOL)hout
{
	return hout;
}

-(int)familyCode
{
	return familyCode;
}

-(NSString*)script
{
	return script;
}

-(int)lod
{
	return lod;
}

-(BOOL)enabled
{
	return enabled;
}

-(BOOL)valid
{
	if([[NSFileManager defaultManager]fileExistsAtPath:script])
	{
		return TRUE;
	}
	else
	{
		return FALSE;
	}
}

-(BOOL)modeSelected
{
	int tmp = (hpmode||nmode)?1:0;
	[self setModeSelected:tmp];
	return modeSelected;
}
//}}}

#pragma mark KVC sets
//{{{
-(void)setName:(NSString*)var
{
	[name release];
	name = [var retain];
}

-(void)setnMode:(bool)var
{
	nmode = var;
	[self setFamilyCode];
}

-(void)sethpMode:(bool)var
{
	hpmode = var;
	[self setFamilyCode];
}

-(void)setMute:(bool)var
{
	mute = var;
	[self setFamilyCode];
}

-(void)setUnmute:(bool)var
{
	unmute = var;
	[self setFamilyCode];
}

-(void)setHin:(bool)var
{
	hin = var;
	[self setFamilyCode];
}

-(void)setHout:(bool)var
{
	hout = var;
	[self setFamilyCode];
}

// note that due to poor programming, the family code 
-(void)setFamilyCode
{	
	int tmp = 0;
	
	if (hout) { tmp++; tmp<<=1; }
	else tmp<<=1;
	if (hin) { tmp++; tmp<<=1; }
	else tmp<<=1;
	if (unmute) { tmp++; tmp<<=1; }
	else tmp<<=1;
	if (mute) { tmp++; tmp<<=1; }
	else tmp<<=1;
	if (hpmode) { tmp++; tmp<<=1; }
	else tmp<<=1;
	if (nmode) { tmp++; tmp<<=1; }
	else tmp<<=1;
	if (enabled) { tmp++; }
	//else tmp<<=1;
	familyCode = tmp;
}

-(void)setModeSelected:(BOOL)var
{
	modeSelected = var;
}

-(void)setScript:(NSString*)var
{
	// If we pass a null string, just run through our stuff, as we probably just want to reset up our scripts
	if(var)
	{
		[script release];
		script = [var retain];
	}

    if(applescript) [applescript release];
	
	// If the script we are given is a path
	if([[NSFileManager defaultManager]fileExistsAtPath:script])
	{
		applescript = [[NSAppleScript alloc] initWithContentsOfURL:[NSURL fileURLWithPath: script] error:nil];
		[self setValid:TRUE];
		[self compile];
	}
	else
	{
		[self setEnabled:FALSE];
		[self setValid:FALSE];
	}
}

-(void)setLod:(int)var
{
	lod = var;
}

-(void)setEnabled:(BOOL)var
{	
	if([[NSFileManager defaultManager]fileExistsAtPath:script])
	{
		[self setValid:TRUE];
		enabled = var;
		[self setFamilyCode];
	}
	else
	{
		enabled = FALSE;
		[self setValid:FALSE];
		[self setFamilyCode];
	}
}

-(void)setValid:(BOOL)var
{
	valid = var;
}

//}}}

-(NSDictionary*)export
{
	//NSLog(@"%@,%i,%i,%i,%@, %i",name,mode,trigger,lod,script,enabled);
	NSDictionary* tmpDict = [NSDictionary dictionaryWithObjectsAndKeys:
		name, @"name",
		[NSNumber numberWithBool:nmode], @"nmode",
		[NSNumber numberWithBool:hpmode], @"hpmode",
		[NSNumber numberWithBool:mute], @"mute",
		[NSNumber numberWithBool:unmute], @"unmute",
		[NSNumber numberWithBool:hin], @"hin",
		[NSNumber numberWithBool:hout], @"hout",
		[NSNumber numberWithInt:lod], @"lod",
		script,@"script",
		[NSNumber numberWithBool:enabled], @"enabled",
		nil];
	return tmpDict;
}

-(void)dealloc
{
    if (applescript) [applescript release];
    if (name) [name release];
    if (script) [script release];
    [super dealloc];
}
@end
